src/catalog/catalog.cpp

```diff
@@ -819,6 +819,42 @@ void Catalog::AddBuiltinFunction(
   function::BuiltInFunctions::AddFunction(func_name, func);
 }
 
+/* @brief
+ * Add a new plpgsql function. This proceeds in two steps:
+ *   1. Add the function information into pg_catalog.pg_proc
+ *   2. Register the function code_context in function::PlgsqlFunction
+ * @param   name & argument_types   function name and arg types used in SQL
+ * @param   return_type   the return type
+ * @param   prolang       the oid of which language the function is
+ * @param   func_src      the plpgsql UDF function body
+ * @details func_src can be used to reconstruct the llvm code_context in case
+ *          of failures
+ * @param   code_context  the code_context that holds the generated LLVM
+ *                        query code
+ */
+void Catalog::AddPlpgsqlFunction(
+    const std::string &name, const std::vector<type::TypeId> &argument_types,
+    const type::TypeId return_type, oid_t prolang, const std::string &func_src,
+    std::shared_ptr<peloton::codegen::CodeContext> code_context,
+    concurrency::TransactionContext *txn) {
+  // Check if UDF already exists
+  auto proc_catalog_obj =
+      ProcCatalog::GetInstance().GetProcByName(name, argument_types, txn);
+
+  if (proc_catalog_obj == nullptr) {
+    if (!ProcCatalog::GetInstance().InsertProc(name, return_type,
```

Comment:
Why is this check needed? Can't you add the 'plpgsql' language on startup?

Reply:
This is not for adding the plpgsql language. This is the ProcCatalogObject returned for the queried UDF name and parameter types. 

This check basically achieves the functionality of checking whether the UDF has already been registered in the catalog or not, and if it has, doesn't not try to register the UDF again.